### PR TITLE
Add wrapper around mcbits.

### PR DESCRIPTION
**Pull request by aparent**
So far I have only added McBits and it is somewhat WIP since the build fails due to libnacl not being available in travis (I put in a request to add it). Am working on added QCBits to this PR.  

Here is the output on my computer:

```
Operation                           Iterations  Total time (s)  Time(us): mean      pop. stdev    Cycles: mean      pop. stdev
RLWE BCNS15
alice 0                                   1223           1.001         818.179         120.504         2855639          420307
bob                                        745           1.001        1343.710         161.471         4690175          563012
alice 1                                   5529           1.000         180.894          46.900          630501          163390
RLWE NewHope
alice 0                                  15431           1.000          64.806          20.136          225484           70086
bob                                      10546           1.000          94.829          36.666          330243          127688
alice 1                                  43433           1.000          23.024          23.353           79751           81467
RLWE MSR LN16
alice 0                                  14046           1.000          71.197          24.621          247755           85758
bob                                       8460           1.000         118.206          24.499          411824           85222
alice 1                                  51963           1.000          19.245          10.685           66474           35554
LWE Frodo recommended
alice 0                                    843           1.001        1186.964         256.613         4142658          895293
bob                                        638           1.000        1567.431         264.284         5470865          922022
alice 1                                   6329           1.000         158.011          28.684          550746           99852
SIDH CLN16
alice 0                                     10           1.088      108800.500        2994.287       379932828        10456107
bob                                          5           1.203      240591.000        3626.731       840153430        12664257
alice 1                                     10           1.089      108873.800        5329.338       380187846        18610019
Code Mcbits
alice 0                                     12           1.009       84119.500       44439.387       293744816       155184482
bob                                      29364           1.000          34.056          22.875          117972           79386
alice 1                                   7754           1.000         128.972          47.746          449156          166384
```

### CODE REVIEW

**Comment by aparent** [comment]
Output with qcbits:

```
Operation                           Iterations  Total time (s)  Time(us): mean      pop. stdev    Cycles: mean      pop. stdev
RLWE BCNS15
alice 0                                   1156           1.001         865.557         141.480         3020635          493552
bob                                        699           1.000        1431.246         206.617         4994752          720938
alice 1                                   5882           1.000         170.031          28.140          592748           97905
RLWE NewHope
alice 0                                  13731           1.000          72.830          19.252          253371           66948
bob                                      10925           1.000          91.539          21.762          318842           75741
alice 1                                  41092           1.000          24.336           9.339           84263           32457
RLWE MSR LN16
alice 0                                  14347           1.000          69.706          15.454          242564           53674
bob                                       8665           1.000         115.419          27.345          402163           95234
alice 1                                  52033           1.000          19.219           8.777           66432           30482
LWE Frodo recommended
alice 0                                    787           1.000        1270.896         261.421         4434965          912462
bob                                        672           1.001        1489.567         235.826         5199324          822959
alice 1                                   6619           1.000         151.087          35.825          526666          124890
SIDH CLN16
alice 0                                     10           1.076      107583.000        5196.044       375680450        18144744
bob                                          4           1.045      261162.750        3776.733       911991074        13188524
alice 1                                     10           1.015      101483.500        3306.069       354380129        11544915
Code Mcbits
alice 0                                     10           1.214      121449.600      105753.074       424104323       369296159
bob                                      27727           1.000          36.067          14.080          125056           48944
alice 1                                   8192           1.000         122.084          29.112          425273          101287
Code QcBits
alice 0                                     82           1.008       12288.598         913.460        42908449         3188537
bob                                      10425           1.000          95.927          23.992          334034           83377
alice 1                                   1046           1.000         956.348         177.110         3337375          617822
```

**Comment by dstebila** [comment]
Thanks for working on this.  Let me know when you'd like me to code review.

Do you know if QcBits is subject to the same attack that Sebastian's QC-MDPC code was?

**Comment by aparent** [comment]
QcBits appears to be an implementation of QC-MDPC so it is likely affected?

The QcBits paper mentions design choices were made with the goal of achieving a low decoding failure rate which probably helps against the attack.  I can look into it.

For now I could modify this pull request so that it only includes McBits so that it can get merged more quickly.

**Comment by dstebila** [comment]
Let's just merge McBits for now.  Once we can confirm QcBits is not affected or has taken countermeasures, then we can merge that too.

**Comment by aparent** [comment]
It is just McBits now. Travis still fails because of the libnacl dependency.  Ideally it would just check to see if libnacl was available then compile with McBits if it is.  Should I wait for the new build system for this?

**Comment by dstebila** [comment]
For now let's just add a flag to our simple Makefile.  Either a boolean flag ENABLE_CODE_MCBITS or a path variable NACL_DIR=whatever.

**Comment by aparent** [comment]
I made the McBits compilation optional. Use `make ENABLE_CODE_MCBITS=1` to enable it. You must have `libnacl` in your include/lib path to compile this way.

**Comment by dstebila** [comment]
I've made a bunch of changes so that McBits will build successfully on macOS when nacl is installed using brew, and configured travis to do that.

Note however, the code doesn't pass our .travis-global-namespace-check.sh, would you be able to take a look at that Alex?

~~~
0000000000000000 T _benes_compact
00000000000005a0 T _bm
0000000000000000 T _into_vec
0000000000000480 T _mask_leq_64bit
0000000000000460 T _mask_nonzero_64bit
00000000000004a0 T _vec_cmov
00000000000000c0 T _vec_reduce
0000000000000000 T _acc
00000000000002e0 T _decrypt
0000000000000270 T _syndrome_adjust
0000000000000080 T _weight
00000000000019c0 T _encrypt
0000000000000000 T _gen_e
0000000000000320 T _syndrome
0000000000000140 T _butterflies
0000000000000490 T _fft
0000000000000000 T _radix_conversions
0000000000000210 T _butterflies_tr
0000000000003b90 T _fft_tr
0000000000000000 T _radix_conversions_tr
0000000000000930 T _GF_mul
0000000000000910 T _gf_diff
0000000000000130 T _gf_inv
0000000000000000 T _gf_mul
00000000000000c0 T _gf_sq
0000000000000000 T _crypto_encrypt
00000000000001b0 T _crypto_encrypt_keypair
00000000000000e0 T _crypto_encrypt_open
0000000000000000 T _pk_gen
0000000000000000 T _irr_gen
0000000000000380 T _sk_gen
0000000000000000 T _transpose_64x64_compact
0000000000000420 T _transpose_8x64
0000000000000050 T _load8
0000000000000000 T _store8
0000000000000ab0 T _vec_copy
0000000000000b60 T _vec_inv
0000000000000000 T _vec_mul
0000000000000b20 T _vec_or
00000000000009e0 T _vec_sq
~~~

**Comment by dstebila** [comment]
@christianpaquin When you get a chance, would you be able to look at updating the VS project file on this branch, since we've added a new directory and build files for McBits?  One complication is that this has a build dependency on external library (https://nacl.cr.yp.to/) -- I am not sure whether nacl even builds for Windows, and if so how to link to it.

**Comment by christianpaquin** [comment]
I'll take a look.

**Comment by dstebila** [comment]
Sadly, with brew all the files are dumped directly in /usr/local/include, no nacl subdirectory.  Maybe we'll have to do some kind of macro to distinguish the two cases?

File: src/kex_code_mcbits/external/sk_gen.c [hunk 1]
```diff
@@ -1,6 +1,6 @@
 #include "sk_gen.h"
 
-#include <nacl/randombytes.h>
+#include <randombytes.h>
```
Comment (dstebila) [inline src/kex_code_mcbits/external/sk_gen.c]: Sadly, with brew all the files are dumped directly in /usr/local/include, no nacl subdirectory. Maybe we'll have to do some kind of macro to distinguish the two cases?

File: src/kex/kex.c [hunk 2]
```diff
@@ -6,6 +6,7 @@
 #include <oqs/kex_rlwe_msrln16.h>
 #include <oqs/kex_lwe_frodo.h>
 #include <oqs/kex_sidh_cln16.h>
+#include <oqs/kex_code_mcbits.h>
```
Comment (christianpaquin) [inline src/kex/kex.c]: This should be wrapped with a #ifdef ENABLE_CODE_MCBITS. The Visual Studio project doesn't currently copy "kex_code_mcbits.h" into the "oqs" dir; this line therefore breaks the Windows build.

**Comment by christianpaquin** [comment]
I looked into building this with Visual Studio. It isn't trivial to build NaCl on Windows. I suggest that this gets committed without initial support on Windows; maybe I'll be able to add it later. I've made one comment (see above) on the kex.c changes to avoid breaking the Windows build.

**Comment by aparent** [comment]
@christianpaquin Do you know if libsodium is easier to work with in windows? It also includes these functions I think.

**Comment by aparent** [comment]
The symbol name spacing issue comes from linking with McBits.  I could change it to build `mcbits.a` outside of `liboqs.a` and just link it in to produce `test_kex`?

I think we should probably do it that way;  especially once we link in NTRU since we don't want to modify that code. It means we cant really guarantee there wont be namespace collisions once everything is linked together though.

**Comment by dstebila** [comment]
I do think we ought to try to guarantee that there won't be namespace collisions once everything is linked together.  Among other reasons, it seems quite likely that we will some day add another implementation that uses a supercop-like API, and we would definitely have a collision with crypto_encrypt/crypto_encrypt_keypair/crypto_encrypt_open.  Renaming can be a huge pain, although an IDE should do it fairly easily. Alternatively is there anything we can make static and use via includes rather than linking?

**Comment by christianpaquin** [comment]
I just tried libsodium and it indeed compiles nicely on Windows, so the switch is a good idea.

**Comment by aparent** [comment]
Great! So I think this is just blocked on the symbols issue which I will fix today.

**Pull request merged**